prototyping/auto-segmentation/sb/05-mini-data-take-2/unet-lung-2d-bn-and-wbce.ipynb

```diff
@@ -0,0 +1,518 @@
+    "output_channels = int(mask_dims[-1])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "ZzufytjWJLRf"
+   },
+   "outputs": [],
+   "source": [
+    "inputs = tf.keras.layers.Input((grid_size, grid_size, 1))\n",
+    "x = inputs\n",
+    "skips = []\n",
+    "\n",
+    "for number_of_filters in [32, 64, 128]:\n",
+    "    x, skip = encode(x, number_of_filters)\n",
+    "    skips.append(skip)\n",
+    "    \n",
+    "skips.reverse()\n",
+    "\n",
+    "for number_of_filters, skip in zip([256, 128, 64], skips):\n",
+    "    x = decode(x, skip, number_of_filters)\n",
+    "    \n",
+    "x = tf.keras.layers.Conv2D(\n",
+    "    output_channels,\n",
+    "    1,\n",
+    "    activation=\"sigmoid\",\n",
+    "    padding=\"same\",\n",
+    "    kernel_initializer=\"he_normal\",\n",
+    ")(x)\n",
+    "\n",
+    "model = tf.keras.Model(inputs=inputs, outputs=x)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "bcnWXRGmJLRf",
+    "outputId": "6f5f482f-ea46-4f16-9116-ae0b1b9ed976"
+   },
+   "outputs": [],
+   "source": [
+    "model.summary()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "RtGmTtDh9xy_",
+    "outputId": "58b14746-c7b4-489d-d2b0-4fb6c3b28cb1"
+   },
+   "outputs": [],
+   "source": [
+    "balancing = len(np.ravel(training_masks)) / np.sum(training_masks) - 1\n",
+    "balancing"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "xGGSXFCi9s5x"
+   },
+   "outputs": [],
+   "source": [
+    "def weighted_bce(y_true, y_pred):\n",
+    "  weights = (y_true * balancing) + 1.\n",
+    "  bce = K.binary_crossentropy(y_true, y_pred)\n",
+    "  loss = K.mean(bce * weights)\n",
+    "  return loss"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 313
+    },
+    "id": "EHYE0T1VJLRf",
+    "outputId": "b03c382a-866c-49ac-a6fd-8a759698ee3a"
+   },
+   "outputs": [],
+   "source": [
+    "model.compile(\n",
+    "    optimizer=tf.keras.optimizers.Adam(),\n",
+    "    loss=weighted_bce,\n",
+    "    metrics=[\n",
+    "        tf.keras.metrics.BinaryAccuracy(),\n",
+    "        tf.keras.metrics.Recall(),\n",
+    "        tf.keras.metrics.Precision()\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "display(sample_image, sample_mask)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 1000
+    },
+    "id": "0TNPQy5AJLRg",
+    "outputId": "b5235583-0a18-432c-e493-48a569349519"
+   },
+   "outputs": [],
+   "source": [
+    "history = model.fit(\n",
```

Comment:
If this is for demonstration purposes, we can also test if the model has enough degrees of freedom to overfit a subset of our data.

```python
sub_index = 10
overfit_test = model.fit(
    training_images[:sub_index], 
    training_masks[:sub_index],
    epochs=500,
    validation_data=(validation_images, validation_masks)
```